executor/load_data.go

```diff
@@ -132,6 +186,111 @@ type FieldMapping struct {
 	UserVar *ast.VariableExpr
 }
 
+// Load reads from readerFn and do load data job.
+func (e *LoadDataInfo) Load(ctx context.Context, readerFn func() ([]byte, error)) error {
+	e.InitQueues()
+	e.SetMaxRowsInBatch(uint64(e.Ctx.GetSessionVars().DMLBatchSize))
+	e.StartStopWatcher()
+	// let stop watcher goroutine quit
+	defer e.ForceQuit()
+	err := sessiontxn.NewTxn(ctx, e.Ctx)
+	if err != nil {
+		return err
+	}
+	// processStream process input data, enqueue commit task
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go processStream(ctx, readerFn, e, wg)
+	err = e.CommitWork(ctx)
+	wg.Wait()
+	return err
+}
+
+// processStream process input stream from network
+func processStream(ctx context.Context, readerFn func() ([]byte, error), loadDataInfo *LoadDataInfo, wg *sync.WaitGroup) {
+	var err error
+	var shouldBreak bool
+	var prevData, curData []byte
+	defer func() {
+		r := recover()
+		if r != nil {
+			logutil.Logger(ctx).Error("process routine panicked",
+				zap.Reflect("r", r),
+				zap.Stack("stack"))
+		}
+		if err != nil || r != nil {
+			loadDataInfo.ForceQuit()
+		} else {
+			loadDataInfo.CloseTaskQueue()
+		}
+		wg.Done()
+	}()
+	for {
+		curData, err = readerFn()
+		if err != nil {
+			if terror.ErrorNotEqual(err, io.EOF) {
+				logutil.Logger(ctx).Error("read data for LOAD DATA failed", zap.Error(err))
+				break
+			}
+			err = nil
```

Comment:
What is the difference? Line 231 has checked non-EOF error, here `err` should be always EOF?

Reply:
yes, I want to clear the EOF error. I guess some readerFn will return (empty curData, nil error) then (empty curData, EOF error), other readerFn will return (non-empty curData, EOF error). I want to finish the task only when get empty curData so EOF error is reset here. If I don't reset it, it will goes to "load data process stream error" branch.